### DISCUSSION TITLE
Invalid Operation Encountered: The Operation received was either Invalid or Corrupted, please retry!

### DISCUSSION THREAD

**Discussion by anonymous9915** [0xTheProDev/Operational-Transformation#110] [Bug Report]
Describe the bug
If the history is not empty, and then if I make any changes then it is throwing an error and also changes are not getting saved to the firebase.
Uncaught (in promise): Invalid Operation Encountered: The Operation received was either Invalid or Corrupted, please retry!
Invalid Operation Encountered: The Operation received was either Invalid or Corrupted, please retry!

The base length of the second operation has to be the target length of the first operation
Assertion Failed: The base length of the second operation has to be the target length of the first operation

Retain expects a positive integer.
Assertion Failed: Retain expects a positive integer.

To Reproduce
Create the firepad using FireMonacoEditor. then write something to the editor and refresh the page, you will get one of those errors.

Expected behavior
Data should be saved to the firebase without throwing errors.

Screenshots
![image](https://user-images.githubusercontent.com/87244355/173988111-5c2e0c4a-876f-420d-a8f8-da52376ee590.png)

Desktop:

OS: Windows
Browser chrome
Version: 102.0.5005.115 (Official Build) (64-bit)
Framework: Angular 12

**Comment by 0xTheProDev** [comment]
Hi @anonymous9915,

Since you are using Windows, my quick educative guess would be this is something to do with line ending which windows just does not get right. Can you once run the example project and see if you can reproduce the same?

**Reply by stefanbugge** [reply]
I can also reproduce this error if one user is on windows and another on osx, so it seems that line endings could be the cause of the error.

**Reply by stefanbugge** [reply]
If you force the monaco editor to a specific line ending the error goes away.

```
editor.getModel().setEOL(0); // 0 is for line feed LF.
```

**Reply by cj-rajat** [reply]
Yes, I did the same to fix.
But I don't know why these errors are still coming in production for many of our users irrespective of the operation system.
sharing some of the errors below ->
Assertion Failed: Retain expects a positive integer.
Validation Failed: Expected a Truth value
Invalid Operation Encountered: The Operation recieved was either Invalid or Corrupted, please retry!
Error: The base length of the second operation has to be the target length of the first operation

**Reply by stefanbugge** [reply]
> Yes, I did the same to fix. But I don't know why these errors are still coming in production for many of our users irrespective of the operation system. sharing some of the errors below -> Assertion Failed: Retain expects a positive integer. Validation Failed: Expected a Truth value Invalid Operation Encountered: The Operation recieved was either Invalid or Corrupted, please retry! Error: The base length of the second operation has to be the target length of the first operation

The problem will persist if you have any stored models in firebase that have another EOL character than the one you set the monaco editor to use. You will have to migrate those stored models as well.

**Reply by 0xTheProDev** [reply]
@stefanbugge It is so awesome to see you take up and help the community. A big thanks to you. 🥳
